[Ong Hong Xun Keith] iP

keithxun opened the pull request:
### Bopes Chatbot

>"This bot is so easy to use." - No One Ever

Bopes frees your legs to track down and execute ~~people~~ tasks to obtain ~~rewards~~ happiness.
It is the single source of truth in life (Source - [Trust me bro](https://github.com/keithxun))

It is,
- Not in vim
- But in java
- And worth 4 MC

To set it up, you just need to do the following:
1. **Download** and **Run** the exe file
2. ~~No skill issue~~ 
3. Profit

And it only took me half my **LIFESPAN**!

Features:
- [x] Completed task
- [ ] Incomplete task
- [ ] Procastinate (Too lazy to be implmented)

---

If you are a Java lover 🤮🤮🤮, you will enjoy this instrument. Here is some `inline code`:
```java
public class Main {
    public static void main(String[] args) {
        new Bopes("data/tasks.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

keithxun committed:
Replace Duke with Bopes as new name for the project

keithxun committed:
Add echo functionality

keithxun committed:
Add list functionality for user inputs

keithxun committed:
Add task class and completion status

keithxun committed:
Add 3 types of tasks: Deadlines, Events, ToDos

keithxun committed:
Add testing functionality

keithxun committed:
Add error handling

keithxun committed:
Add delete functionality

keithxun committed:
Improve testing and ui

keithxun committed:
Add file storage functionality

keithxun committed:
Merge branch-Level-7 into master

keithxun committed:
Add Date and Time to Deadline and Event

keithxun committed:
Merge branch-Level-8 into master

keithxun committed:
Improve OOP

keithxun committed:
Improve ui

keithxun committed:
Merge branch-A-MoreOOP into master

keithxun committed:
Add packages

keithxun committed:
Remove duplicate ui for marking task

keithxun committed:
Fix isDone bug when loading from file

keithxun committed:
Merge branch-A-Package into master

keithxun committed:
Merge remote-tracking branch 'remotes/origin/add-gradle-support'

keithxun committed:
Add gradle

keithxun committed:
Add test cases and fix bugs

keithxun committed:
Config for jar file

keithxun committed:
Add javadoc for exception

keithxun committed:
Add Javadoc for Deadline

keithxun committed:
Add javadoc for Event

keithxun committed:
Add javadoc for Task

keithxun committed:
Add javadoc to TaskList

keithxun committed:
Add javadoc for ToDo

keithxun committed:
Add Javadoc for Parser

keithxun committed:
Add javadoc to storage

keithxun committed:
Add javadoc to Ui

keithxun committed:
Fix identation for switch statements

keithxun committed:
Add find functionality

keithxun committed:
Merge branch-A-JavaDoc into master

keithxun committed:
Merge branch-A-CodingStandard into master

keithxun committed:
Merge branch-Level-9 into master

wenjebs commented on src/main/java/bopes/Parser.java:
```diff
@@ -0,0 +1,187 @@
+        String[] commandWords = fullCommand.split(" ", 2);
+        String commandType = commandWords[0];
+
+        switch (commandType) {
+            case "find":
+                handleFindCommand(commandWords[1], tasks, ui);
+                break;
+            case "list":
+                ui.showTasks(tasks);
+                break;
+            case "mark":
+                handleMarkCommand(commandWords[1], tasks, ui, storage);
+                break;
+            case "unmark":
+                handleUnmarkCommand(commandWords[1], tasks, ui, storage);
+                break;
+            case "delete":
+                handleDeleteCommand(commandWords[1], tasks, ui, storage);
+                break;
+            default:
+                handleAddTaskCommand(fullCommand, tasks, ui, storage);
+                break;
```
Need to align with the switch case style guide?

keithxun replied:
Resolved.

wenjebs commented on src/main/java/bopes/exception/BopesException.java:
```diff
@@ -0,0 +1,63 @@
+    public static BopesException invalidIndex(int size) {
+        return new BopesException("Error: The task index is out of range. Please provide a valid task number between 1 and " + size + ".");
+    }
+
+    /**
+     * Returns a BopesException indicating that the user input is not a valid number.
+     *
+     * @return a BopesException with a message indicating that the input is not a valid number
+     */
+    public static BopesException invalidNumberFormat() {
+        return new BopesException("Error: The provided input is not a valid number. Please enter a valid task number.");
+    }
+
+    /**
+     * Returns a BopesException indicating that the deadline format provided by the user is incorrect.
+     *
+     * @return a BopesException with a message indicating the correct deadline format
+     */
+    public static BopesException invalidDeadlineFormat() {
+        return new BopesException("Error: Incorrect deadline format. Please use the format 'deadline <task> /by <dd/MM/yyyy hh:mm a>'.");
+    }
+
+    /**
+     * Returns a BopesException indicating that the event format provided by the user is incorrect.
+     *
+     * @return a BopesException with a message indicating the correct event format
+     */
+    public static BopesException invalidEventFormat() {
+        return new BopesException("Error: Incorrect event format. Please use the format 'event <task> /from <dd/MM/yyyy hh:mm a> /to <dd/MM/yyyy hh:mm a>'.");
+    }
+
+    /**
+     * Returns a BopesException indicating that the user has entered an unknown command.
+     *
+     * @return a BopesException with a message indicating the valid commands and their usage
+     */
+    public static BopesException unknownCommand() {
+        return new BopesException("Error: Unknown command. Please use 'todo', 'deadline', or 'event' followed by the task details. Example: 'todo read book'. Use 'bye' to exit the program.");
+    }
+}
```
I like how you did this !

wenjebs reviewed (changes requested):
Looks good! 
Requested one switch statement styling

keithxun committed:
Fix switch indentation

keithxun committed:
Add CheckStyle configuration files

keithxun committed:
Fix styling for test files

keithxun committed:
Fix CheckStyle issues with Bopes class

keithxun committed:
Fix CheckStyle issues with BopesException class

keithxun committed:
Fix bug in Deadline and Event classes regarding DateTime parsing

keithxun committed:
Working GUI with DialogBox and MainWindow

keithxun committed:
Fix list functionality

keithxun committed:
Fix error messages

keithxun committed:
Merge branch 'branch-Level-10'

kaoxi998533 commented on src/main/java/bopes/DialogBox.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * DialogBox represents a dialog box containing a speaker's message and avatar.
+ */
+public class DialogBox extends HBox {
```
I like how you have very understandable name for this class

kaoxi998533 commented on src/main/java/bopes/exception/BopesException.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return a BopesException with a message indicating the correct deadline format
+     */
+    public static BopesException invalidDeadlineFormat() {
```
I like how you have names lucidly explaining the different exception instances that the static method returns

kaoxi998533 commented on src/main/java/bopes/task/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+ * The Deadline class represents a task with a specific deadline.
+ * It extends the Task class and includes a LocalDateTime field to store the deadline date and time.
+ */
+public class Deadline extends Task {
```
Should this be DeadlineTask instead, since Deadline is not really a task if interpreted without context

kaoxi998533 commented on src/main/java/bopes/Storage.java:
```diff
@@ -0,0 +1,77 @@
+     * @param tasks the TaskList containing the tasks to be saved
+     * @throws BopesException if there is an error saving tasks to the file
+     */
+    public void saveTasks(TaskList tasks) throws BopesException {
```
I like how this method name clearly shows what the method does

kaoxi998533 reviewed (commented):
LGTM!
Overall the naming conventions are followed!

keithxun committed:
Add assertions for deadline class

keithxun committed:
Add assertions to Event class

keithxun committed:
Add assertions to TaskList and ToDo classes

keithxun committed:
Add assertions for the parser

keithxun committed:
Refactor parse method in parser

keithxun committed:
Refactor addTask method in parser

keithxun committed:
Add streams to find command in TaskList

keithxun committed:
Merge pull request #1 from keithxun/branch-A-Assertions

Branch a assertions

keithxun committed:
Merge branch 'master' into branch-A-CodeQuality

keithxun committed:
Merge branch 'master' into branch-A-Streams

keithxun committed:
Merge pull request #2 from keithxun/branch-A-CodeQuality

Branch a code quality

keithxun committed:
Merge branch 'master' into branch-A-Streams

keithxun committed:
Merge pull request #4 from keithxun/branch-A-Streams

Branch a streams

keithxun committed:
Add functionality to sort tasks by date and time

keithxun committed:
Move the error catching from the Bopes class to the Parser class

Currently error handling is done directly in the Bopes class and all
its component propogate the error message up to the Bopes class.
This is not ideal as it makes the code harder to read and maintain.
It would be better to move the error handling to the Parser class
as it is responsible for parsing the user input.

This will help to make the code more modular and easier to read.
The Parser class is responsible for parsing the user input,
so it makes sense for it to handle the error catching as well.

With this change, the Bopes class will only need to call the Parser
and it will also be easier to debug.

keithxun committed:
Merge pull request #6 from keithxun/branch-C-Sort

Branch c sort

keithxun committed:
Add MainWindow responsive layout

keithxun committed:
Add styles for dialog box

keithxun committed:
Add different styling for error messages

keithxun committed:
Merge branch 'branch-A-BetterGui'

keithxun committed:
Add ai.md

keithxun committed:
Add user guide

keithxun committed:
Add margins

keithxun committed:
Update product screenshot

keithxun committed:
Update storage file path

keithxun committed:
Change coding style of main classes

keithxun committed:
Change coding style for tasks

keithxun committed:
Fix all checkstyle comments

keithxun committed:
Update user guide

keithxun committed:
Add javadocs for test files

keithxun committed:
Fix storage format for Deadline and Event tasks

keithxun committed:
Add launcher class

keithxun committed:
Remove duplicate "Error: " prefix in error messages
